### ISSUE TITLE
[REVIEW]: FAT Forensics: A Python Toolbox for Implementing and Deploying Fairness, Accountability and Transparency Algorithms in Predictive Systems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1904]
**Submitting author:** @So-Cool (<a href="http://orcid.org/0000-0002-9869-5896">Kacper Sokol</a>)
**Repository:** <a href="https://github.com/fat-forensics/fat-forensics" target ="_blank">https://github.com/fat-forensics/fat-forensics</a>
**Version:** 0.0.1
**Editor:** @arokem
**Reviewer:** @bernease, @osolari
**Archive:** <a href="https://doi.org/10.5281/zenodo.3833199" target="_blank">10.5281/zenodo.3833199</a>

## Status

[![status](https://joss.theoj.org/papers/070c8b6b705bb47d1432673a1eb03f0c/status.svg)](https://joss.theoj.org/papers/070c8b6b705bb47d1432673a1eb03f0c)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/070c8b6b705bb47d1432673a1eb03f0c"><img src="https://joss.theoj.org/papers/070c8b6b705bb47d1432673a1eb03f0c/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/070c8b6b705bb47d1432673a1eb03f0c/status.svg)](https://joss.theoj.org/papers/070c8b6b705bb47d1432673a1eb03f0c)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/fat-forensics/fat-forensics" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@bernease & @osolari, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @arokem know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fat-forensics/fat-forensics">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@So-Cool) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fat-forensics/fat-forensics">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@So-Cool) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @bernease, @ @osolari it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1904 with the following error: 

 pandoc-citeproc: reference osolari not found
Error producing PDF.
! TeX capacity exceeded, sorry [input stack size=5000].
\reserved@a ->\def \reserved@a 
                               *{\let \@xs@assign \@xs@expand@and@detokenize...
l.324 }

Looks like we failed to compile the PDF

**Comment by danielskatz** [comment]
👋 @arokem - I can't quite understand the status of this review - can you let me know where it is?

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1904 with the following error: 

 pandoc-citeproc: reference osolari not found
Error producing PDF.
! TeX capacity exceeded, sorry [input stack size=5000].
\reserved@a ->\def \reserved@a 
                               *{\let \@xs@assign \@xs@expand@and@detokenize...
l.324 }

Looks like we failed to compile the PDF

**Comment by danielskatz** [comment]
👋 @arfon - something seems messed up here - the reference that pandoc mentions isn't in the paper, I don't think, but is one of the reviewers (?!)  And the reviewer info in the first comment in this issue seems to have an extra ```@``` - Any ideas?

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01904/joss.01904/10.21105.joss.01904.pdf)

**Comment by arfon** [comment]
> 👋 @arfon - something seems messed up here - the reference that pandoc mentions isn't in the paper, I don't think, but is one of the reviewers (?!) And the reviewer info in the first comment in this issue seems to have an extra @ - Any ideas?

Very weird. I'm not sure what's going on here. I've manually fixed up the review issue at the top. Also, the compilation error is related to the extra `@` which was then being fed into Pandoc (as we now include reviewer names in the compiled papers).

**Comment by danielskatz** [comment]
Thanks @arfon

**Comment by danielskatz** [comment]
👋 @arokem - this is now ready for you to start the review process

**Comment by arokem** [comment]
@bernease, @osolari: where does this review currently stand? Have you had a chance to take a look?

**Comment by osolari** [comment]
I am currently finishing my review.

**Comment by osolari** [comment]
Hi, I wasn't able to install the package. Did others manage to install it using simple pip?

**Comment by osolari** [comment]
@whedon @arokem I finished my review, but could not install the package using pip. Have you given it a try?

**Comment by osolari** [comment]
@bernease I finished my review, but could not install the package using pip. Have you given it a try?

**Comment by So-Cool** [comment]
@osolari, have you tried `pip install fat-forensics`?

**Comment by osolari** [comment]
yes!

**Comment by So-Cool** [comment]
@osolari, could you please paste here the log of executing this command?

**Comment by osolari** [comment]
ERROR: Could not find a version that satisfies the requirement fat-forensics (from versions: none)
ERROR: No matching distribution found for fat-forensics

**Comment by osolari** [comment]
nvm. I tried it on my own machine and it was installed. sorry!

**Comment by So-Cool** [comment]
Are you trying to install it on Windows? At the moment only Linux and MacOS are supported. You may try to install it directly from sources (which may or may not work):
```
pip install git+https://github.com/fat-forensics/fat-forensics.git
```

**Comment by osolari** [comment]
both machines were Macs. Solved thanks! 

Btw, should I just ping @whedon  when I'm done?

**Comment by osolari** [comment]
@whedon @arfon @arokem I'm done with the review!

**Comment by arokem** [comment]
Hi @osolari,  Thanks for your review! I see that there are still two boxes left unchecked in your review (functionality and references). Is this because you still have issues with these two items? If these are not properly addressed right now, could you please make some comments about the changes you believe need to be made?

**Comment by osolari** [comment]
Sorry I forgot to check functionality. 

About the other two though. I think a brief literature review of the field with more references would help me understand the importance of the work better. Currently the report only references FATFORENSIC itself.

**Comment by osolari** [comment]
*by check I meant "check off".

**Comment by osolari** [comment]
@arokem I was wondering if I would receive an email indicating that I have completed my review service. :-)

**Comment by arokem** [comment]
Hey @osolari : have your previously-raised concerns ([here](https://github.com/openjournals/joss-reviews/issues/1904#issuecomment-569320507)) been addressed?

**Comment by osolari** [comment]
@arokem no it hasn't. :-)

**Comment by arokem** [comment]
Hey @bernease : have you had a chance to take a look at this? Thanks!

**Comment by osolari** [comment]
@arokem Would someone please confirm that I have submitted the review?

**Comment by arokem** [comment]
Hi @osolari : I believe that the authors are currently still waiting for comments from @bernease to proceed with their revision. I will ping her by email to see when she might have the chance to take a look.

**Comment by bernease** [comment]
Sorry for letting this fall off my radar. I can have the feedback into you by EOD Thursday conservatively, perhaps earlier if I'm able to clear some time later today.

**Comment by bernease** [comment]
I have completed the review to the best of my ability.

I have some concerns about the lack of references to related work as well as discussion of related software in the field, leading me keep these two fields unchecked for now. There is a singular reference to a pre-print manuscript describing the FATForensics software package in depth, Sokol, Santos-Rodriguez, and Flach (2019). 

At minimum, would like to see directly comparative software packages in Python (e.g., eli5, AI Fairness 360) referenced and compared to FATForensics in JOSS submission. As the article stands, it is difficult to see where this work fits into the landscape of software without consulting the other manuscript from the authors. Should that be sufficient, I will mark these two criteria as met.

**Comment by So-Cool** [comment]
Thank you for the review, @bernease. I'll address your and @osolari's comments over the weekend and update the paper. I'll make sure to let you know when this is done.

**Comment by So-Cool** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01904/joss.01904/10.21105.joss.01904.pdf)

**Comment by So-Cool** [comment]
We revised the paper taking into consideration your feedback and comments. @osolari and @bernease, please let us know if you find something that we can improve.

**Comment by bernease** [comment]
I have updated (and completed all marks) based on the updates provided by the author.

**Comment by arokem** [comment]
@whedon check DOI

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by arokem** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a  dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by arokem** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.5281/zenodo.1423046 is OK

MISSING DOIs

- https://doi.org/10.18653/v1/n16-3020 may be missing for title: "Why Should I Trust You?": Explaining the Predictions of Any Classifier
- https://doi.org/10.1080/10618600.2014.907095 may be missing for title: Peeking inside the black box: Visualizing statistical learning with plots of individual conditional expectation

INVALID DOIs

- None
```

**Comment by arokem** [comment]
@So-Cool : could you please add the DOIs mentioned above? 

Two other comments about references: 

1. Since scikit learn is mentioned so prominently, it would be appropriate to cite their 2011 JMLR paper about the software: http://www.jmlr.org/papers/v12/pedregosa11a.html

2. What is the difference between this paper and the arXiv preprint referenced? Is that preprint already published elsewhere? 

Thanks!

**Comment by So-Cool** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.18653/v1/n16-3020 is OK
- 10.1214/aos/1013203451 is OK
- 10.1080/10618600.2014.907095 is OK
- 10.5281/zenodo.1423046 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by So-Cool** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01904/joss.01904/10.21105.joss.01904.pdf)

**Comment by So-Cool** [comment]
Hi @arokem,
I added the missing DOIs and referenced scikit-learn; thank you for the pointers. The arXiv preprint is an extended version of the JOSS paper with usage examples and more thorough discussion of the relevant literature. We will be revising it in the coming months and submitting it to either a journal or a conference.

If any other revisions are needed, just let me know.

**Comment by arokem** [comment]
Thank you. Before we proceed, could you please:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g. figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by arokem** [comment]
Hi @So-Cool : have you had a chance to complete these steps? Any questions?

**Comment by So-Cool** [comment]
Hi @arokem, I'm merging two more PRs before creating the release, hence the delay. I should wrap it up soon. I hope that's OK with you.

**Comment by arokem** [comment]
Yup. Just let me know here when it settles.

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by arfon** [comment]
:wave: @So-Cool, just a friendly check-in to see how things are going with those last few steps outlined by @arokem?

**Comment by So-Cool** [comment]
Thanks @arfon. We've had a bit of a delay but the things are heading in the right direction now and we should wrap it up relatively soon.

**Comment by So-Cool** [comment]
Hi @arfon and @arokem,
apologies for the delay. The release tag is [`0.1.0`](https://github.com/fat-forensics/fat-forensics/releases/tag/0.1.0). The DOI of the release is [10.5281/zenodo.3833199](https://doi.org/10.5281/zenodo.3833199).

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.3833199 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3833199" target="_blank">10.5281/zenodo.3833199</a> is the archive.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01904/joss.01904/10.21105.joss.01904.pdf)

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.18653/v1/n16-3020 is OK
- 10.1214/aos/1013203451 is OK
- 10.1080/10618600.2014.907095 is OK
- 10.5281/zenodo.1423046 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1452

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1452, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1262696721521532929) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1453
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01904](https://doi.org/10.21105/joss.01904)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@bernease, @osolari - many thanks for your reviews here and to @arokem for editing this submission ✨ 

@So-Cool - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01904/status.svg)](https://doi.org/10.21105/joss.01904)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01904">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01904/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01904/status.svg
   :target: https://doi.org/10.21105/joss.01904
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01904/status.svg)](https://doi.org/10.21105/joss.01904)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
